[inductor] Fix reference_as_float gradcheck

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106581
* #106747
* #106870
* __->__ #106626

When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

peterbell10 committed:
[inductor] Fix reference_as_float gradcheck

When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Fix reference_as_float gradcheck"


When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -386,9 +390,16 @@ def run(*ex, **kwargs):
             self.assertEqual(
-                actual_grad,
-                correct_grad,
+                flat_actual,
+                flat_expect,
```
In the forward they unflatten it before passing it to assertEqual, perhasp you should do here the same?

peterbell10 committed:
Update on "[inductor] Fix reference_as_float gradcheck"


When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -386,9 +389,15 @@ def run(*ex, **kwargs):
             actual_grad = compute_grads(example_inputs, kwargs, actual, grads)
+
+            if reference_in_float:
+                expect_grad = reference_to_expect(actual_grad, correct_grad)
```
why is it correct not to flatten/unflatten here? Don't the gradients have the same structure as the forward operation?

peterbell10 replied:
`compute_grads` returns from `torch.autograd.grad(...)` which is always a flat tuple according to the [docstring](https://pytorch.org/docs/stable/generated/torch.autograd.grad.html).

peterbell10 committed:
rebase on "[inductor] Fix reference_as_float gradcheck"


When `reference_as_float` is true, reference gradients will not have the same
dtype as the actual computed gradients. This fixes the issue by downcasting
before doing the comparison.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

lezcano reviewed (approved):
Fair enough
